synapse/storage/databases/main/events_worker.py

```diff
@@ -2072,29 +2072,29 @@ async def get_already_persisted_events(
         """
 
         mapping = {}
-        txn_id_to_event: Dict[Tuple[str, int, str], str] = {}
+        txn_id_to_event: Dict[Tuple[str, str, str], str] = {}
 
         for event in events:
-            token_id = getattr(event.internal_metadata, "token_id", None)
+            device_id = getattr(event.internal_metadata, "device_id", None)
```

Comment:
Since device IDs are user-controlled, is this safe? I thought the pair (user ID, device ID) was the right (read: refresh-token-aware) replacement for a token.

Reply:
Good catch. 👍